cmd/server/main.go

```diff
@@ -17,31 +27,146 @@ func main() {
+	defer func() {
+		logger.G.Info("Cancel context")
+		cancel()
+	}()
+	// Регистрируем прослушиватель для закрытия записи в файл и завершения сервера
+	go func() {
+		stop := make(chan os.Signal, 1)
+		signal.Notify(stop, syscall.SIGINT, syscall.SIGTERM)
+		<-stop
+		logger.G.Info("Stopping server")
+		cancel()
+	}()
+
+	InitLogger(func() {
+		InitStore(func() {
+			if err := run(ctx, &wg); err != nil { // Запускаем сервер
+				logger.G.Error(err)
+			}
+		}, ctx, &wg)
+	}, ctx)
 }
 
 // run запуск сервера
-func run() error {
-	log.Println("Running server on", config.Params.Address)
-	return http.ListenAndServe(config.Params.Address, getRouter())
+func run(ctx context.Context, wg *sync.WaitGroup) (err error) {
+	logger.G.Infof("Running server on %s", config.Params.Address)
+	server := http.Server{
+		Addr:    config.Params.Address,
+		Handler: getRouter(),
+	}
+	wg.Add(1)
+	go func() {
+		err = stopServer(&server, ctx, wg)
+	}()
+	err = server.ListenAndServe()
+	if err != nil && !errors.Is(err, http.ErrServerClosed) {
+		return err
+	}
+	return nil
+}
+
+func stopServer(server *http.Server, ctx context.Context, wg *sync.WaitGroup) error {
+	<-ctx.Done()
+	// Заставляем завершиться сервер и ждём его завершения
+	err := server.Shutdown(ctx)
+	if err != nil {
+		logger.G.Errorf("Failed to shutdown server: %v", err)
+	}
+	logger.G.Info("Server stop")
+	wg.Done()
+	return err
 }
 
 // getRouter конфигурация роутинга приложение
 func getRouter() chi.Router {
 	router := chi.NewRouter()
-	// Сохранение метрики
-	router.Post("/update/{type}/{name}/{value}", handlemetric.Handler)
+	// Устанавилваем мидлваре
+	router.Use(
+		cMiddleware.StripSlashes,          // Убираем лишние слеши
+		logger.LogRequests,                // Логируем данные запроса
+		middlewares.GZIPCompressResponse,  // Сжимаем ответ TODO исключить для роутов, которые будут возвращать не application/json или text/html. Проверять в мидлваре или компрессоре может быть не эффективно,так как заголовок с контентом может быть поставлен позже записи контента
+		logger.LogResponse,                // Логируем данные ответа
+		middlewares.GZIPDecompressRequest, // Разжимаем тело ответа
+	)
+	// Сохранение метрики по URL
+	router.Post("/update/{type}/{name}/{value}", handlemetric.URLHandler)
 	// Получение всех метрик
 	router.Get("/", getmetrics.Handler)
 	// Получение отдельной метрики
-	router.Get("/value/{type}/{name}", getmetric.Handler)
+	router.Get("/value/{type}/{name}", getmetric.URLHandler)
+
+	router.Group(func(r chi.Router) {
+		// Устанавилваем мидлваре с логированием запросов
+		r.Use(middlewares.JSONHeaders)
+		// Сохранение метрики с помощью JSON тела
+		r.Post("/update", handlemetric.JSONHandler)
+		// Получение отдельной метрики
+		r.Post("/value", getmetric.JSONHandler)
+	})
 	return router
 }
+
+type next func()
+
+// InitStore устанавливаем глобальное хранилище метрик.
+func InitStore(n next, ctx context.Context, wg *sync.WaitGroup) {
```

Comment:
По конвенции Go, контекст передается как первый параметр функции.

```suggestion
func InitStore(ctx context.Context, n next,  wg *sync.WaitGroup) {
```

Reply:
Перенёс и упростил функции